causalpy/exp_inverse_propensity_weighting.py

```diff
@@ -0,0 +1,433 @@
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Inverse propensity weighting
+"""
+
+import arviz as az
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from matplotlib.lines import Line2D
+from patsy import dmatrices
+from sklearn.linear_model import LinearRegression as sk_lin_reg
+
+from causalpy.data_validation import PropensityDataValidator
+from causalpy.experiments import ExperimentalDesign
+
+# from causalpy.pymc_models import PyMCModel
+# from causalpy.utils import round_num
+
+
+class InversePropensityWeighting(ExperimentalDesign, PropensityDataValidator):
+    """
+    A class to analyse inverse propensity weighting experiments.
+
+    :param data:
+        A pandas dataframe
+    :param formula:
+        A statistical model formula for the propensity model
+    :param outcome_variable
+        A string denoting the outcome variable in datq to be reweighted
+    :param weighting_scheme:
+        A string denoting which weighting scheme to use among: 'raw', 'robust',
+        'doubly robust' or 'overlap'. See Aronow and Miller "Foundations
+        of Agnostic Statistics" for discussion and computation of these
+        weighting schemes.
+    :param model:
+        A PyMC model
+
+    Example
+    --------
+    >>> import causalpy as cp
+    >>> df = cp.load_data("nhefs")
+    >>> seed = 42
+    >>> result = cp.InversePropensityWeighting(
+    ...     df,
+    ...     formula="trt ~ 1 + age + race",
+    ...     outcome_variable ="outcome",
+    ...     weighting_scheme="robust",
+    ...     model=cp.pymc_models.PropensityScore(
+    ...         sample_kwargs={
+    ...             "draws": 100,
+    ...             "target_accept": 0.95,
+    ...             "random_seed": seed,
+    ...             "progressbar": False,
+    ...         },
+    ...     ),
+    ... )
+    """
+
+    def __init__(
+        self,
+        data: pd.DataFrame,
+        formula: str,
+        outcome_variable: str,
+        weighting_scheme: str,
+        model=None,
+        **kwargs,
+    ):
+        super().__init__(model=model, **kwargs)
+        self.expt_type = "Inverse Propensity Score Weighting"
+        self.data = data
+        self.formula = formula
+        self.outcome_variable = outcome_variable
+        self.weighting_scheme = weighting_scheme
+        self._input_validation()
+
+        t, X = dmatrices(formula, self.data)
+        self._t_design_info = t.design_info
+        self._t_design_info = X.design_info
+        self.labels = X.design_info.column_names
+        self.t, self.X = np.asarray(t), np.asarray(X)
+        self.y = self.data[self.outcome_variable]
+
+        COORDS = {"obs_ind": list(range(self.X.shape[0])), "coeffs": self.labels}
+        self.coords = COORDS
+        self.model.fit(X=self.X, t=self.t, coords=COORDS)
+
+    def make_robust_adjustments(self, ps):
```

Comment:
Even though we ar not using MyPy yet. For this refactor I suggest to use return types as well :)

Reply:
I've now added type hints for returns for all plotting functions